lib/backend/firestore/firestorebk.go

```diff
@@ -396,8 +426,21 @@ func (b *Backend) GetRange(ctx context.Context, startKey []byte, endKey []byte,
 		}
 
 		if r.isExpired(b.clock.Now()) {
-			if _, err := docSnap.Ref.Delete(ctx); err != nil {
-				return nil, ConvertGRPCError(err)
+			if _, err := docSnap.Ref.Delete(ctx, firestore.LastUpdateTime(docSnap.UpdateTime)); err != nil && status.Code(err) == codes.FailedPrecondition {
+				// If the document has been updated, then attempt one additional get to see if the
+				// resource was updated and is no longer expired.
+				docSnap, err := b.svc.Collection(b.CollectionName).Doc(docSnap.Ref.ID).Get(ctx)
+				if err != nil {
+					return nil, ConvertGRPCError(err)
+				}
+				r, err := newRecordFromDoc(docSnap)
+				if err != nil {
+					return nil, trace.Wrap(err)
+				}
+
+				if !r.isExpired(b.clock.Now()) {
+					values = append(values, r.backendItem())
+				}
```

Comment:
low-priority nit: this doesn't violate the backend model, but there may be cases where it would result in what I'd call "annoying" behavior:

Say that I have some range `R`,  and the following sequence of operations happens:

- An initial range request occurs across range `R` at time `T1`.
- A write happens to key `R1` at time `T2`.
- A write happens to key `R2` at time `T3`.
- A delete of the old value of `R2`  is attempted by the original range request, falling back to a read of `R2` at time `T4`.

In this sequence of events, the resulting range contains the value of `R2` as set at `T3`, but omits the value of `R1` as set at `T2`.

This is ultimately _fine_, and can happen for any number of legitimate reasons, since many of our backends either don't guarantee that writes to different keys are well-ordered in the event stream, and/or because range requests are evaluated sequentially from the "latest" state, so values that appear later in the range may also be "newer".

However, all else being equal, less inconsistencies of this nature are probably better, and in this case I think we could have less network round-trips, simpler logic, and more consistency by just partially or completely omitting this logic.

We already treat expiry as a lazy/best-effort affair, and given the limitations of our backends we're going to keep doing that at least to some extent.  I'd suggest either completely omitting optimistic cleanup, or simply returning the expired value if optimistic cleanup fails for any reason.

Additionally, it might be worth putting a per-request cap on the amount of optimistic cleanup we're willing to do... if a cluster was offline for a while, we might start up to a backend state with 20_000 expired nodes. We don't really want a `GetRange` call waiting on 20_000 API requests I think.  Better to just let those 20_000 nodes get whittled down asynchronously.

Anyhow, like I said... not actually a violation of the backend model, just a potential annoyance, so take it or leave it.